src/coreclr/interpreter/compiler.cpp

```diff
@@ -2303,25 +2301,9 @@ bool InterpCompiler::EmitNamedIntrinsicCall(NamedIntrinsic ni, CORINFO_CLASS_HAN
         case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReferenceOrContainsReferences:
         {
             CORINFO_CLASS_HANDLE clsHnd = sig.sigInst.methInst[0];
-            bool isValueType = (m_compHnd->getClassAttribs(clsHnd) & CORINFO_FLG_VALUECLASS) != 0;
-            bool hasGCRefs = false;
-
-            if (isValueType)
-            {
-                // Walk the layout to see if any field is a GC pointer
-                const uint32_t maxSlots = 256;
-                BYTE gcLayout[maxSlots];
-                uint32_t numSlots = m_compHnd->getClassGClayout(clsHnd, gcLayout);
-
-                for (uint32_t i = 0; i < numSlots; ++i)
-                {
-                    if (gcLayout[i] != TYPE_GC_NONE)
-                    {
-                        hasGCRefs = true;
-                        break;
-                    }
-                }
-            }
+            uint32_t clsAttribs = m_compHnd->getClassAttribs(clsHnd);
+            bool isValueType = (clsAttribs & CORINFO_FLG_VALUECLASS) != 0;
+            bool hasGCRefs = (clsAttribs & CORINFO_FLG_CONTAINS_GC_PTR) != 0;
```

Comment:
Does this produce correct result for byref-like types, e.g. `RuntimeHelpers.IsReferenceOrContainsReferences<Span<int>>()`?

Reply:
So this doesn't work only for byref types ? In this case you should also update the usage in `InterpCompiler::AllocOffsets` since it means we have a bug there.

Reply:
Can I do that in a follow-up PR?

Reply:
sure